module/fastqc.nf

```diff
@@ -0,0 +1,45 @@
+/*
+*   Nextflow module for running FASTQC
+*
+*   @input fq_path path path to the input FASTQ file
+*   @output fastqc_output_dir dir unzipped FASTQC output directory
+*/
+
+include { generate_standard_filename } from '../external/pipeline-Nextflow-module/modules/common/generate_standardized_filename/main.nf'
+
+process assess_ReadQuality_FastQC {
+    container params.docker_image_fastqc
+
+    publishDir path: "${params.workflow_output_dir}/output",
+        pattern: "${sm_id}/${output_filename}_fastqc",
+        mode: "copy",
+        enabled: true
+    ext log_dir_suffix: { "-${target}" }
+
+    input:
+        tuple path(path), val(unused), val(sm_id), val(rg_arg), val(rg_id), val(unused), val(unused), val(ununsed)
+
+    output:
+        path("${sm_id}/${output_filename}_fastqc")
+
+    script:
+    target = "${sm_id}-${rg_id}"
+    output_filename = generate_standard_filename("FastQC-${params.fastqc_version}",
+        params.dataset_id,
+        target,
+        [:])
+
+    """
+    set -euo pipefail
+    mkdir -p ${sm_id}
+    samtools view -F 0x900 -h ${rg_arg} ${path} | \
+        samtools fastq | \
+        fastqc \
+        --outdir "${sm_id}" \
+        --format fastq \
+        --extract \
+        --delete \
+        ${params.fastqc_additional_options} \
+        stdin:${output_filename}
```

Comment:
@sorelfitzgibbon 
**question:** I think I understand what this command tries to archive but can you elaborate on this a bit (can this `-F 0x900` be optional, etc)? Also, have you tried to incorporate threading? This would be probably really slow for large samples. Is it possible to use `fastqc -format bam` with threading for the sample level run?

Reply:
Thanks @tyamaguchi-ucla 
- `-F 0x900` removes secondary and supplementary alignments.  It is actually the default but I thought I'd leave it here so there is no question that its being done.
- Surprisingly `fastqc -format bam` does not allow per readgroup analysis.
- I have not tried to incorporate threading.  All three commands allow threading.   I will try it and test speedup.

Reply:
Below are the results for running with various threads on `SGSALASC000015-T011-O04-F.bam` (185 GB).  
- This sample has five readgroups with one being much larger than the others.
- All output here: `/hot/software/pipeline/pipeline-generate-SQC-BAM/Nextflow/development/unreleased/sfitz-by-readgroup`
- It looks like the sweet spot is 2 threads
#### Speedup (single thread realtime/realtime)
threads | 1 | 2 | 3 | 4 | 6 | 12
|--|--|--|--|--|--|--|
|small readgroup|1 |1.9|2.1|3.8|4|3.4|
|large readgroup|1|2|2.5|2.2|2.6|2.5|
|total cpu hours|9.3|9.2|12.2|12.4|16.7|37.0|



#### 1 thread; full run CPU HOURS = 9.3
name|threads|realtime|%cpu|peak_rss|peak_vmem|rchar|wchar
--|--|--|--|--|--|--|--
assess_ReadQuality_FastQC (2)|1|1h 24m 33s|97.50%|444.2 MB|3.3 GB|369.5 GB|185.2 GB 
assess_ReadQuality_FastQC (5)|1|1h 27m 31s|99.70%|444.8 MB|3.3 GB|369.6 GB|185.4 GB 
assess_ReadQuality_FastQC (4)|1|1h 27m 31s|99.70%|447.5 MB|3.3 GB|369.6 GB|185.4 GB 
assess_ReadQuality_FastQC (3)|1|1h 27m 31s|99.80%|445.4 MB|3.3 GB|370.4 GB|186.2 GB 
assess_ReadQuality_FastQC (1)|1|3h 31m 22s|99.60%|444.5 MB|3.3 GB|1.2 TB|1 TB

#### 2 threads; full run cpu hours = 9.2
name|threads|realtime|%cpu|peak_rss|peak_vmem|rchar|wchar
--|--|--|--|--|--|--|--
assess_ReadQuality_FastQC (3)|2|44m 11s|188.30%|459.8 MB|3.9 GB|370.4 GB|186.2 GB
assess_ReadQuality_FastQC (4)|2|44m 11s|190.20%|465.7 MB|3.9 GB|369.6 GB|185.4 GB
assess_ReadQuality_FastQC (5)|2|44m 10s|189.90%|465.2 MB|3.9 GB|369.6 GB|185.4 GB
assess_ReadQuality_FastQC (2)|2|38m 54s|191.90%|462 MB|3.9 GB|369.5 GB|185.2 GB
assess_ReadQuality_FastQC (1)|2|1h 45m 35s|200.20%|459.1 MB|3.9 GB|1.2 TB|1 TB

#### 3 threads; full run cpu hours = 12.2
name|threads|realtime|%cpu|peak_rss|peak_vmem|rchar|wchar
--|--|--|--|--|--|--|--
assess_ReadQuality_FastQC (2)|3|40m 48s|182.00%|877.1 MB|4.9 GB|369.5 GB|185.2 GB
assess_ReadQuality_FastQC (3)|3|40m 48s|181.90%|885.5 MB|4.9 GB|370.4 GB|186.2 GB
assess_ReadQuality_FastQC (4)|3|38m 42s|187.30%|885.5 MB|4.9 GB|369.6 GB|185.4 GB
assess_ReadQuality_FastQC (5)|3|38m 43s|187.50%|880.5 MB|5 GB|369.6 GB|185.4 GB
assess_ReadQuality_FastQC (1)|3|1h 24m 36s|237.80%|894.7 MB|5 GB|1.2 TB|1 TB

#### 4 threads; full run cpu hours = 12.4
name|threads|realtime|%cpu|peak_rss|peak_vmem|rchar|wchar
--|--|--|--|--|--|--|--
assess_ReadQuality_FastQC (2)|4|22m 33s|350.10%|1.1 GB|5.7 GB|369.5 GB|185.2 GB
assess_ReadQuality_FastQC (3)|4|22m 19s|358.50%|905.9 MB|5.8 GB|370.4 GB|186.2 GB
assess_ReadQuality_FastQC (5)|4|22m 7s|356.40%|970.9 MB|5.7 GB|369.6 GB|185.4 GB
assess_ReadQuality_FastQC (4)|4|22m 10s|353.60%|971.2 MB|5.7 GB|369.6 GB|185.4 GB
assess_ReadQuality_FastQC (1)|4|1h 37m 8s|233.10%|1.1 GB|5.7 GB|1.2 TB|1 TB 

#### 6 threads; full run CPU HOURS = 16.7
name|threads|realtime|%cpu|peak_rss|peak_vmem|rchar|wchar
--|--|--|--|--|--|--|--
assess_ReadQuality_FastQC (2)|6|21m 36s|332.00%|1.5 GB|6.9 GB|369.5 GB|185.2 GB
assess_ReadQuality_FastQC (3)|6|21m 36s|319.90%|1.4 GB|6.8 GB|370.4 GB|186.2 GB
assess_ReadQuality_FastQC (5)|6|21m 36s|332.20%|1.5 GB|6.9 GB|369.6 GB|185.4 GB
assess_ReadQuality_FastQC (4)|6|21m 36s|318.70%|1.5 GB|6.9 GB|369.6 GB|185.4 GB
assess_ReadQuality_FastQC (1)|6|1h 20m 29s|263.50%|1.4 GB|7 GB|1.2 TB|1 TB

#### 12 threads; full run CPU HOURS = 37.0
name|threads|realtime|%cpu|peak_rss|peak_vmem|rchar|wchar
--|--|--|--|--|--|--|--
assess_ReadQuality_FastQC (5)|12|25m 22s|271.30%|2.5 GB|10.4 GB|369.6 GB|185.4 GB
assess_ReadQuality_FastQC (2)|12|25m 22s|265.00%|2.3 GB|10.6 GB|369.5 GB|185.2 GB
assess_ReadQuality_FastQC (4)|12|25m 22s|265.00%|2.4 GB|10.5 GB|369.6 GB|185.4 GB
assess_ReadQuality_FastQC (3)|12|25m 22s|271.30%|2.5 GB|10.4 GB|370.4 GB|186.2 GB
assess_ReadQuality_FastQC (1)|12|1h 23m 44s|250.30%|2.1 GB|10.5 GB|1.2 TB|1 TB

Reply:
> -F 0x900 removes secondary and supplementary alignments. It is actually the default but I thought I'd leave it here so there is no question that its being done.

Thanks for the benchmarking results. I wasn't aware of this default. Do we want to include them though (i.e. `-f`?) to get a complete result for the RG level FASTQs, etc?

Reply:
No, I don't think we do want to include them.  They are duplicates of reads that are already represented.  Removing them creates the output that directly matches the original FASTQs.

Reply:
Right. That's generally the case. Also, the `-f $flag` won't work as it specifically extracts reads with the flag, I think.

Back in Canada, we encountered some issues where certain (older) BAMs didn't work as described and we had to specifically include secondary alignments (I think that was with picard) for back conversion. I guess we can't really do much here if the input has issues.

Reply:
@sorelfitzgibbon (**suggestion/question: non-blocking**) You've generated lots of benchmarking results and we should post them on GitHub discussions as well. Also, do you typically adjust for caching performance when conducting benchmarking?

Reply:
> do you typically adjust for caching performance when conducting benchmarking

Is there a recommended way of doing that?  I ran almost all of it on nodes that were already running and had space available.  They were run at more or less the same time.  I could run them all concurrently on a single F72, although I assume even that background will change as jobs start to finish.

Reply:
Caching adjustments are hard to control in general; a way around it could to be make the tests totally independent of the cache by copying the input files to /scratch and then run all benchmarking runs using files under scratch.

Reply:
I will do some cache independent benchmarking.

Reply:
> Caching adjustments are hard to control in general; a way around it could to be make the tests totally independent of the cache by copying the input files to /scratch and then run all benchmarking runs using files under scratch.

Yeah, I was going suggest this. Maybe worth creating a confluence page for general benchmark recommendations? We've discussed pre-caching, etc with OHIA and MSFT a while ago but using /scratch would be the easiest solution.